File: Sep/Day18/design-task-manager.cc
#include <bits/stdc++.h>
using namespace std;

class TaskManager {
    unordered_map<int, pair<int,int>> taskMap; // taskId -> (priority, userId)
    priority_queue<pair<pair<int,int>, int>> globalHeap; 
    // stores { {priority, taskId}, taskId } with negative for max-heap effect

public:
    TaskManager(vector<vector<int>>& tasks) {
        for (auto &t : tasks) {
            add(t[0], t[1], t[2]);
        }
    }

    void add(int userId, int taskId, int priority) {
        taskMap[taskId] = {priority, userId};
        globalHeap.push({{priority, taskId}, taskId});
    }

    void edit(int taskId, int newPriority) {
        if (taskMap.find(taskId) != taskMap.end()) {
            int userId = taskMap[taskId].second;
            taskMap[taskId] = {newPriority, userId};
            globalHeap.push({{newPriority, taskId}, taskId});
        }
    }

    void rmv(int taskId) {
        taskMap.erase(taskId);
    }

    int execTop() {
        while (!globalHeap.empty()) {
            auto top = globalHeap.top(); globalHeap.pop();
            int priority = top.first.first;
            int taskId = top.first.second;
            if (taskMap.count(taskId) && taskMap[taskId].first == priority) {
                int userId = taskMap[taskId].second;
                rmv(taskId);
                return userId;
            }
        }
        return -1;
    }
};

File: Sep/Day18/design-task-manager.js
class TaskManager {
    constructor(tasks) {
        this.taskMap = new Map(); // taskId -> [priority, userId]
        this.globalHeap = []; // max-heap simulated with array + sort

        for (let [userId, taskId, priority] of tasks) {
            this.add(userId, taskId, priority);
        }
    }

    add(userId, taskId, priority) {
        this.taskMap.set(taskId, [priority, userId]);
        this.globalHeap.push([priority, taskId]);
        this.globalHeap.sort((a, b) => b[0] - a[0] || b[1] - a[1]); // keep max-heap order
    }

    edit(taskId, newPriority) {
        if (this.taskMap.has(taskId)) {
            let [, userId] = this.taskMap.get(taskId);
            this.taskMap.set(taskId, [newPriority, userId]);
            this.globalHeap.push([newPriority, taskId]);
            this.globalHeap.sort((a, b) => b[0] - a[0] || b[1] - a[1]);
        }
    }

    rmv(taskId) {
        this.taskMap.delete(taskId);
    }

    execTop() {
        while (this.globalHeap.length > 0) {
            let [priority, taskId] = this.globalHeap.shift();
            if (this.taskMap.has(taskId) && this.taskMap.get(taskId)[0] === priority) {
                let [, userId] = this.taskMap.get(taskId);
                this.rmv(taskId);
                return userId;
            }
        }
        return -1;
    }
}
